### DISCUSSION TITLE
Result with table name or not

### DISCUSSION THREAD

**Discussion by cpdivers** [sqlalchemy/sqlalchemy#7855] [Usage Questions] [answered]
Hello everyone,

I've been banging my head on this issue, I looked at the documentation, try google, etc ... but I cannot figure it out.

Here is a test I'm running I have those 2 queries: the first one has a join, the other do not.
I'm trying to have a consistent result with the 2 queries.

```
# With Join
query1 = db.session.query(MinifigInventory,BLColor).join(BLColor, MinifigInventory.COLOR == BLColor.color_id).filter( MinifigInventory.ITEMID == '10124', MinifigInventory.COLOR >= 5)
res1 = query1.all()

# Without Join
query2 = db.session.query(MinifigInventory).filter( MinifigInventory.ITEMID == '10124')
res2 = query2.all()

print(str(query1.statement.compile(dialect=mysql.dialect(), compile_kwargs={"literal_binds": True})))
print(str(query2.statement.compile(dialect=mysql.dialect(), compile_kwargs={"literal_binds": True})))
```

Those queries works, but the behavior and results are different

If I want to print the result of the first query, I have to mention the tablename:
```
for row in res1:
    print(f"{row.MinifigInventory.ITEMID=}")
    print(f"{row.BLColor.color_name=}")
```

Result is:
```
row.MinifigInventory.ITEMID='10124'
row.BLColor.color_name='Dark Tan'
row.MinifigInventory.ITEMID='10124'
row.BLColor.color_name='Medium Azure'
row.MinifigInventory.ITEMID='10124'
row.BLColor.color_name='Light Bluish Gray'
row.MinifigInventory.ITEMID='10124'
```
if I do a ```print(res1)``` it returns a list of object such as: ```[(<MinifigInventory 26470>, <BLColor 69>), (<MinifigInventory 29164>, <BLColor 156>), (<MinifigInventory 32904>, <BLColor 86>), ......]```

What I would really like to achieve is to return all column of the MinifigInventory table + BLColor.name column of the BLColor table, and all of this without mentioning any table

Now the second query, when running a ```print(res2)``` it returns this ```[<MinifigInventory 26470>, <MinifigInventory 29164>, ... ]```
so no BLColor table as expected (there is no join)

My 'problem' is that if I want to print the result this loop will fail:
```
for row in res2:
    print(f"{row.MinifigInventory.ITEMID=}")
```

It seems that if only 1 table is returned, the table name is omitted and therefore to fix the 'problem' I have to use this loop, without mentioning the table
```
for x in res2:
        print(f"{x.ITEMID=}")
```

So my questions are:
1) is there a way to modify the query1 so I do not have to mention the table name ? I still need to return 1 column for the BLColor table !
2) is there a way to modify the query2 so I have to mention the table name MinifigInventory ?

I'm trying to have a consistent behavior

Thank you for your help.

**Comment by gordthompson** [comment, marked as answer]
If you're talking about "mention[ing] the table" when retrieving the row values, then have the query return specific elements (columns) instead of ORM objects:

```python
import sqlalchemy as sa
from sqlalchemy.orm import declarative_base, Session

engine = sa.create_engine("sqlite://", echo=True)

Base = declarative_base()


class Thing(Base):
    __tablename__ = "thing"
    id = sa.Column(sa.Integer, primary_key=True, autoincrement=False)
    description = sa.Column(sa.String(50), nullable=False)
    color_id = sa.Column(sa.Integer)


class Color(Base):
    __tablename__ = "color"
    id = sa.Column(sa.Integer, primary_key=True, autoincrement=False)
    name = sa.Column(sa.String(50), nullable=False)


Base.metadata.create_all(engine)

with Session(engine) as sess:
    # example data
    sess.add(Color(id=1, name="red"))
    sess.add(Color(id=2, name="green"))
    sess.add(Thing(id=10124, description="apple", color_id=1))
    sess.add(Thing(id=90210, description="grape", color_id=2))
    sess.commit()

    # test query
    qry = (
        sa.select(Thing.id, Thing.description, Color.name.label("color_name"))
        .select_from(Thing)
        .join(Color, Thing.color_id == Color.id)
    )
    results = sess.execute(qry)
    for row in results:
        print(
            f"id={row.id}, description='{row.description}', color_name='{row.color_name}'"
        )
        """
        id=10124, description='apple', color_name='red'
        id=90210, description='grape', color_name='green'
        """
```

**Reply by cpdivers** [reply]
Thank you, I'm going to give it a shot,

Since I'm new to this SQLAlchemy and also Python, is there a way to achieve this with ORM only ?

I'm also using
```
from flask_sqlalchemy import SQLAlchemy
```

I did start with RAW SQL query, I wonder if I should have stick with it :-) It was much easier for me, but I'm too advanced in my project to go back to it

**Reply by gordthompson** [reply]
> is there a way to achieve this with ORM only ?

My answer *is* "ORM only", although I will concede that sometimes the distinction between ORM and Core can be a bit blurry. Essentially, ORM is about objects and Core is about tables/columns. They're two different ways of looking at the same data model, and SQLAlchemy tries very hard to let you work with your data in the most appropriate fashion for your particular task.

**Reply by cpdivers** [reply]
I believe I'm confused, because the way you do it is I believe called 'CORE' but in your import you use:
```from sqlalchemy.orm import declarative_base, Session```

**Reply by gordthompson** [reply]
SQLAlchemy 1.4 introduced several changes to make using Core and ORM more consistent, so that may be contributing to the confusion. For comparison, here is my sample code above using Core:

```python
import sqlalchemy as sa

engine = sa.create_engine("sqlite://", echo=True)

metadata = sa.MetaData()

thing = sa.Table(
    "thing",
    metadata,
    sa.Column("id", sa.Integer, primary_key=True, autoincrement=False),
    sa.Column("description", sa.String(50), nullable=False),
    sa.Column("color_id", sa.Integer),
)

color = sa.Table(
    "color",
    metadata,
    sa.Column("id", sa.Integer, primary_key=True, autoincrement=False),
    sa.Column("name", sa.String(50), nullable=False),
)

metadata.create_all(engine)

with engine.connect() as conn:
    # example data
    conn.execute(
        color.insert(), [dict(id=1, name="red"), dict(id=2, name="green")]
    )
    conn.execute(
        thing.insert(),
        [
            dict(id=10124, description="apple", color_id=1),
            dict(id=90210, description="grape", color_id=2),
        ],
    )
    conn.commit()

    # test query
    qry = (
        sa.select(
            thing.c.id, thing.c.description, color.c.name.label("color_name")
        )
        .select_from(thing)
        .join(color, thing.c.color_id == color.c.id)
    )
    results = conn.execute(qry)
    for row in results:
        print(
            f"id={row.id}, description='{row.description}', color_name='{row.color_name}'"
        )
        """
        id=10124, description='apple', color_name='red'
        id=90210, description='grape', color_name='green'
        """
```

**Comment by CaselIT** [comment]
> is there a way to modify the query2 so I have to mention the table name MinifigInventory ?

Assuming you are on sqlalchemy 1.4, if instead of `query()` you create a select then execute it on the session you will get this behaviour, since execute always returns row even in case of a single item in them.

You can try
```py
stmt = select(MinifigInventory).filter( MinifigInventory.ITEMID == '10124')
res2 = db.session.execute(stmt).all()
```

**Comment by cpdivers** [comment]
Ok guys, this is going to make me re-thinking a couple things that is for sure.

I know it is a lot to ask, but can you show me how to update my code below ? 
Remember: I'm about 2 weeks old with Python and SQLAlchemy, I still have clearly a lot to learn and fully understand all those imports !!! And I'm sure that at this point I'm mixing up different concept without understanding them fully 


In every case, thanks in advance for your help.


Here is what I have

```
## models.py
from sqlalchemy import ForeignKey, Column, Float, String, TIMESTAMP, Table, Text
from sqlalchemy.sql import text
from sqlalchemy.dialects.mysql import INTEGER, SMALLINT, MEDIUMINT, TINYINT
from sqlalchemy.orm import relationship
from lego2 import db, login_manager
from flask_login import UserMixin


class MinifigInventory(db.Model):
    __tablename__ = 'minifigs_inventory'

    id = Column(INTEGER(11), primary_key=True)
    SETNUMBER = Column(String(20, 'utf8mb4_unicode_ci'), nullable=False, index=True)
    ITEMTYPE = Column(String(1, 'utf8mb4_unicode_ci'), nullable=False)
    ITEMID = Column(String(20, 'utf8mb4_unicode_ci'), nullable=False, index=True)
    QTY = Column(INTEGER(11), nullable=False)
    COLOR = Column(ForeignKey('BL_colors.color_id'), index=True)
    EXTRA = Column(String(1, 'utf8mb4_unicode_ci'))
    ALTERNATE = Column(String(1, 'utf8mb4_unicode_ci'))
    MATCHID = Column(INTEGER(11))
    COUNTERPART = Column(String(1, 'utf8mb4_unicode_ci'))

    BL_color = relationship('BLColor')

class BLColor(db.Model):
    __tablename__ = 'BL_colors'

    color_id = Column(MEDIUMINT(8), primary_key=True, server_default=text("0"))
    color_name = Column(String(30, 'utf8mb4_unicode_ci'), nullable=False, index=True)
    color_rgb = Column(String(10, 'utf8mb4_unicode_ci'))
    color_type = Column(String(20, 'utf8mb4_unicode_ci'), nullable=False)
    COLORCNTPARTS = Column(INTEGER(11), server_default=text("0"))
    COLORCNTSETS = Column(INTEGER(11), server_default=text("0"))
    COLORCNTWANTED = Column(INTEGER(11), server_default=text("0"))
    COLORCNTINV = Column(INTEGER(11), server_default=text("0"))
    COLORYEARFROM = Column(SMALLINT(6), server_default=text("0"))
    COLORYEARTO = Column(SMALLINT(6), server_default=text("0"))
```

```
## __init__.py

from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from flask_bcrypt import Bcrypt
from flaskext.mysql import MySQL  # First database

from flask import json  # use to manage errors
from flask_login import LoginManager

app = Flask(__name__)
app.config["SECRET_KEY"] = "081XX4899XXaef85c40c50##b1e"
app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql://lego:mydbpassword@172.19.32.226/legodb'

app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False            # Remove a warning about deprecation
#app.config["SQLALCHEMY_ECHO"] = True

db = SQLAlchemy(app)

```

```
## routes-test.py

from ast import In
from base64 import urlsafe_b64decode
from unittest import result
from flask import render_template, url_for, redirect, jsonify, abort, flash, request

from formencode import variabledecode
from lego2 import app, bcrypt, db  # , mysql
from lego2.forms import *
from lego2.models import BLColor,  BLMinifig, MinifigInventory
from flask_login import login_user, current_user, logout_user, login_required

from sqlalchemy import *

from sqlalchemy.sql import text, functions
from sqlalchemy.sql.expression import null, outerjoin
from sqlalchemy.orm import load_only
from sqlalchemy.dialects import mysql

import re  # Regex
import requests

@app.route("/test", methods=["POST", "GET"])
def test():
    query1 = db.session.query(MinifigInventory,BLColor).\
            join(BLColor, MinifigInventory.COLOR == BLColor.color_id).\
            filter( MinifigInventory.ITEMID == '10124', MinifigInventory.COLOR >= 5)
    res1 = query1.all()
    
    print(str(query1.statement.compile(dialect=mysql.dialect(), compile_kwargs={"literal_binds": True})))
    print("WITH TABLE NAME")
    print(res1)
    for row in res1:
            print(f"{row.MinifigInventory.ITEMID=}")
            print(f"{row.BLColor.color_name=}")

```

The return for now is:
```
row.MinifigInventory.ITEMID='10124'
row.BLColor.color_name='Olive Green'
row.MinifigInventory.ITEMID='10124'
row.BLColor.color_name='Olive Green'
row.MinifigInventory.ITEMID='10124'
```


But I would like to change my loop to 
```
    print("WITHOUT TABLE NAME")
    print(res1)
    for row in res1:
            print(f"{row.ITEMID=}")
            print(f"{row.color_name=}")
```

And then return this:
``````
row.ITEMID='10124'
row.color_name='Olive Green'
row.ITEMID='10124'
row.color_name='Olive Green'
row.ITEMID='10124'
```

**Reply by gordthompson** [reply]
```python
    query1 = (
        db.session.query(MinifigInventory.ITEMID, BLColor.color_name)
        .join(BLColor, MinifigInventory.COLOR == BLColor.color_id)
        .filter(
            MinifigInventory.ITEMID == "10124", MinifigInventory.COLOR >= 5
        )
    )
```

should work.

Additional notes:

1. `session.query()` is considered a legacy construct. It still works, but new code should use `select()` as in my examples above.
2. `from sqlalchemy import *` - Don't do that. Only import what you need.

**Reply by cpdivers** [reply]
Gord,

I have tried this:
``` 
    query3 = db.session.query(MinifigInventory.ITEMID,MinifigInventory.SETNUMBER,MinifigInventory.QTY,BLColor.color_name).\
            join(BLColor, MinifigInventory.COLOR == BLColor.color_id).\
            filter( MinifigInventory.ITEMID == '10124', MinifigInventory.COLOR >= 5)
    res3 = query3.all()
    
    for row in res3:
            print(f"{row.ITEMID=}")
            print(f"{row.SETNUMBER=}")
            print(f"{row.QTY=}")
            print(f"{row.color_name=}")
```

And it works as expected, expect for the fact the I have to specify every single column I want to return. My tables do not contain that many so it is still manageable, but it might be a pain if the number column if greater.

I'm still a bit disappointed that everything I learn about .query is already obsolete and replaced by .select

**Reply by cpdivers** [reply]
Just tried your 'new' way using a 'select' instead of query:

SELECT
```
    with Session(engine) as sess:
        qry1 = sa.select(MinifigInventory.ITEMID,MinifigInventory.QTY,BLColor.color_name) \
        .select_from(MinifigInventory) \
        .join(BLColor, MinifigInventory.COLOR == BLColor.color_id) \
        .filter(MinifigInventory.ITEMID == '10124', MinifigInventory.COLOR >= 5)
```

QUERY
```
    query3 = db.session.query(MinifigInventory.ITEMID,MinifigInventory.QTY,BLColor.color_name).\
             join(BLColor, MinifigInventory.COLOR == BLColor.color_id).\
             filter( MinifigInventory.ITEMID == '10124', MinifigInventory.COLOR >= 5).all()
```

So in a way it does not change much, except that based on what you say using select is the new way !

Anyhow, my problem is about the same, because I want to return every single column of the table: MinifigInventory, but only the column BLColor.color_name of the joined table.

In everycase, in the 2 queries above, I had to specify every single column I want to return from table MinifigInventory, and 1 column of BLColor. 

Is there a way to return ALL column of MinifigInventory, and add the BLColor.color_name to it ?

Thank again for your help, I'm learning a lot

**Reply by gordthompson** [reply]
`session.execute()` tends to return ORM objects, while `connection.execute()` tends to return rows of individual elements (column values). So we can construct our query by referring to the ORM classes and use a connection `.execute()` to run it:

```python
query1 = (
    sa.select(MinifigInventory, BLColor.color_name)
    .join(BLColor, MinifigInventory.COLOR == BLColor.color_id)
    .filter(MinifigInventory.ITEMID == "10124", MinifigInventory.COLOR >= 5)
)
with engine.connect() as conn:
    res1 = conn.execute(query1).fetchall()
    for row in res1:
        print(row)
        print(f"{row.ITEMID=}")
        print(f"{row.color_name=}")
"""
(1, '10124', 5, 'green')
row.ITEMID='10124'
row.color_name='green'
"""
```

**Reply by cpdivers** [reply]
Gord,

Is there somewhere a document which explain clearly the difference between `session.execute()` and `connection.execute()` ?

You definitely seem to know a couple things or two that I do not :-)

Side question: 

Is there a point or is it even possible to 'translate' this query into ORM ? 

```
SELECT explo.SETNUMBER,
explo.ITEMTYPE,
explo.ITEMID,
explo.QTY,
explo.EXTRA,
explo.ALTERNATE,
explo.MATCHID,
explo.COUNTERPART,
explo.SOURCE,
co.color_id,
co.color_name,
co.color_type,
pa.ITEMDESC,
pa.CATEGORY,
ca.category_name
FROM (
	SELECT * FROM set_and_minifig_inventory WHERE setnumber = '8960-1' 
	UNION
	SELECT * FROM set_and_minifig_inventory WHERE setnumber IN 
		(
		SELECT ITEMID FROM set_and_minifig_inventory WHERE setnumber = '8960-1' AND ITEMTYPE='M')
		) AS explo
INNER JOIN BL_colors AS co ON explo.COLOR = co.color_id
LEFT JOIN BL_parts AS pa ON explo.ITEMID = pa.ITEMID
LEFT JOIN BL_categories AS ca ON ca.category_id = pa.CATEGORY
WHERE explo.ITEMTYPE <> 'M'
ORDER BY explo.SETNUMBER
```

How can I run this RAW SQL Query with SQL Alchemy ?

**Reply by zzzeek** [reply]
> Gord,
> 
> Is there somewhere a document which explain clearly the difference between `session.execute()` and `connection.execute()` ?

i would spend time going through the [unified tutorial](https://docs.sqlalchemy.org/en/14/tutorial/index.html) that goes through great lengths to make all the foundational concepts clear.  For this specific question, the discussion starts at https://docs.sqlalchemy.org/en/14/tutorial/dbapi_transactions.html#tutorial-executing-orm-session .

**Reply by zzzeek** [reply]
then it continues re: session executing at https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#tutorial-selecting-orm-entities

**Reply by gordthompson** [reply]
> How can I run this RAW SQL Query with SQL Alchemy ?

You can wrap it in a [text()](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.text) object and connection.execute() that.
